ibis/expr/datatypes.py

```diff
@@ -1,47 +1,72 @@
+import builtins
 import collections
 import datetime
+import functools
 import itertools
 import numbers
 import re
+import typing
+
+from typing import (
+    Any as GenericAny,
+    Callable,
+    Iterator,
+    List,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Set as GenericSet,
+    Tuple,
+    TypeVar,
+    Union,
+)
 
-import six
 import pandas as pd
 
 import toolz
 from multipledispatch import Dispatcher
 
 import ibis.common as com
-from ibis.compat import PY2, builtins, functools
 import ibis.expr.types as ir
 
 
-class DataType(object):
-
+class DataType:
     __slots__ = 'nullable',
 
-    def __init__(self, nullable=True):
+    def __init__(self, nullable: bool = True) -> None:
         self.nullable = nullable
 
-    def __call__(self, nullable=True):
+    def __call__(self, nullable: bool = True) -> 'DataType':
+        if nullable is not True and nullable is not False:
+            raise TypeError(
+                "__call__ only accepts the 'nullable' argument. "
+                "Please construct a new instance of the type to change the "
+                "values of the attributes."
+            )
         return self._factory(nullable=nullable)
 
-    def _factory(self, nullable=True):
-        return type(self)(nullable=nullable)
+    def _factory(self, nullable: bool = True) -> 'DataType':
+        slots = {
+            slot: getattr(self, slot) for slot in self.__slots__
+            if slot != 'nullable'
+        }
+        return type(self)(nullable=nullable, **slots)
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
         return self.equals(other)
 
-    def __ne__(self, other):
-        return not self.__eq__(other)
+    def __ne__(self, other) -> bool:
+        return not (self == other)
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         custom_parts = tuple(
             getattr(self, slot)
             for slot in toolz.unique(self.__slots__ + ('nullable',))
```

Comment:
I guees `toolz.unique` is not required, and `nullable` is part of `__slots__`.  Same goes for `__repr__`.

Reply:
Since `__slots__` isn't inherited it will only be contained in the `DataType` class. I use unique because I want `nullable` to be part of the hash of child class instances, and I don't want the hash of `DataType` itself to be `hash((self.nullable, self.nullable))`. Admittedly, it is very unlikely that anyone would ever hash an instance of `DataType` itself.